### PR TITLE
Reduce number of requests to an Ethereum client

### PR DESCRIPTION
**Pull request by CodeSandwich**
- Switch from polling for the pool state to listening to the pool events
- The same for Dai token
- Stop querying for contract constants, instead have them hardcoded
- Fix unwatching of the attestation contract

### CODE REVIEW

File: ui/src/funding/contract.ts [hunk 1]
```diff
@@ -87,59 +75,272 @@ export class PoolContract {
...
+    interface StreamFragment {
+      // Timestamp since which a stream sends funds (inclusive)
+      since: number;
+      // Timestamp to which a stream sends funds (exclusively)
+      to: number;
+      amtPerSec: BigNumber;
+    }
+    // For each sender stores a list of non-overlapping stream fragments
+    // in chronological order
+    const streams = new Map<string, StreamFragment[]>();
```
Comment (geigerzaehler) [inline ui/src/funding/contract.ts:211]: This is state that is mutated by events through `onStreamUpdate` and read by `collectableFn`. I think it would make sense to encapsulate this in a class to better separate it from the rest of the code.
```typescript
class Collectable {
  private streams: Map<string, StreamFragment[]>;

  function collectableFn(now: Date): Big
  function onStreamUpdated(event: Event): Promise<void>
}
```

**Review by geigerzaehler — COMMENTED** [review]
Nice, this should be a big improvement. I only have some architectural comments the third commit.

File: ui/src/funding/contract.ts [hunk 2]
```diff
@@ -87,59 +75,272 @@ export class PoolContract {
...
+  // Start watching the state of a given pool sender.
+  // `onUpdated` is called immediately with the latest state.
+  // Returns a function, which unwatches the state when called.
+  async watchPoolSender(
+    sender: string,
+    onUpdated: (data: PoolSenderData) => void
+  ): Promise<() => void> {
+    const int = this.contract.interface;
+    // The event must be SenderUpdated
+    async function getSenderData(event: Event): Promise<PoolSenderData> {
```
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply ui/src/funding/contract.ts:120]: I've tested it out and I'm not sure about the benefits of that setup. I've applied your advice and it indeed worked. But then I've replaced all `SenderUpdatedEventData` fields with nonexistent ones and the compiler hasn't even blinked, everything just kept working fine. It's probably due to the mandatory `extends Array<unknown>`, the ethers API prevents us from static checking anything.

[the remark this answers is not in the document]
Comment (geigerzaehler) [reply ui/src/funding/contract.ts:120]: I think this is caused by `ErcPool` not being typed at all :exploding_head:. The following is possible
```typescript
function explode(pool: Erc20Pool) {
  pool.aslkdfj();
}
```
https://github.com/ethereum-ts/TypeChain/pull/373 fixes it. I’ll update `radicle-contracts`.

File: ui/src/funding/contract.ts (hunk 1 above)
Comment (CodeSandwich) [reply ui/src/funding/contract.ts]: Fixed

File: ui/src/funding/pool.ts [hunk 3]
```diff
@@ -335,3 +311,70 @@ export function isValidBig(value: string): boolean {
...
 }
+
+class PoolWatcher {
+  private pool: contract.PoolContract;
+  private token: daiToken.ERC20;
+  private ethAddr: string;
+  private unwatch: () => void;
+  private data: PoolData;
+  private balanceFn: (now: Date) => Big;
+  private collectableFn: (now: Date) => Big;
```
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply ui/src/funding/pool.ts]: Good point, fixed

File: ui/src/funding/pool.ts [hunk 4]
```diff
@@ -335,3 +311,70 @@ export function isValidBig(value: string): boolean {
...
+
+  constructor(pool: contract.PoolContract, token: daiToken.ERC20) {
+    this.pool = pool;
+    this.token = token;
+    this.ethAddr = "uninitialized";
```
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply ui/src/funding/pool.ts]: Right, it makes the intention clearer, fixed

File: ui/src/funding/contract.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply ui/src/funding/contract.ts]: Fixed

File: ui/src/funding/contract.ts [hunk 5]
```diff
@@ -87,59 +75,272 @@ export class PoolContract {
...
+    ) => {
+      await onCollected(event);
+      // For now it doesn't make sense to call `onUpdated` here,
+      // because `collectableFn` doesn't change.
+      // This is to keep the behavior of the `watchPoolReceiver`
+      // align with its documentation and leave room for addition of
+      // more watched parameters, e.g. the last collection time
+      onUpdated(collectableFn);
+    };
+    this.contract.on(filterCollected, collectedListener);
```
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply ui/src/funding/contract.ts:197]: I'm not sure if it's going to be beneficial. These are mostly pass-throughs, they don't provide much abstraction and remove little boilerplate. They introduce hard to follow signatures and often don't match the actual use cases requiring wrapping arguments and returned closures in additional logic.

The main benefit I can see is batching of `contract.on` and `contract.off`, which is nice, but IMO not worth it.

Also I was unable to pull off the `...args` trick cleanly, I'm forced to give it type `Event[]`, which is a dangerous lie and I need to handle `pop()` returning `undefined`, which will never happen.

File: ui/src/funding/contract.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply ui/src/funding/contract.ts]: Added, it's much better after the upgrade! The new typed events prevent accessing fields absent from the data interface. It doesn't guard us from adding invalid fields to the interface, but that's already a huge safety and maintainability improvement.

File: ui/src/funding/contract.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (geigerzaehler) [reply ui/src/funding/contract.ts:222]: Good points. Given that you extracted all of the logic into separated functions now the listening logic is much easier to follow. I’m ok if we keep it like this.

**Review by geigerzaehler — APPROVED** [review]
Nice. Another big improvement!

**Comment by geigerzaehler** [comment]
I’ve merged https://github.com/radicle-dev/radicle-contracts/pull/141 so you need to update `package.json`

**Comment by CodeSandwich** [comment]
Radicle-contracts updated to the current `master`

**Review by geigerzaehler — APPROVED** [review]
